### PR TITLE
Juggler plugin in go

### PR DESCRIPTION
**Pull request by sakateka**
Juggler plugin rewrite in go,
also this PR introduce new senders data type `[]tasks.AggregationResult`.
All senders take data in it

### CODE REVIEW

File: senders/juggler/api.go [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+	Tags             []string               `json:"tags"`
+	Methods          []string               `json:"methods"`
+	Children         []jugglerChildrenCheck `json:"children"`
+	Flap             *jugglerFlapConfig     `json:"flaps,omitempty"`
+
+	//Active           string                  `json:"active"`
+	//ActiveKWArgs     map[string]string       `json:"active_kwargs"`
+	//AlertInterval    []int64                `json:"alert_interval"`
+	//RefreshTime      int64                  `json:"refresh_time"`
+	//Ttl              int64                  `json:"ttl"`
```
Comment (noxiouz) [inline senders/juggler/api.go]: На будущее?

File: senders/juggler/api.go [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+			check.Service = t.Service
+			if err := js.updateCheck(ctx, check); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (js *Sender) ensureFlap(jcheck *jugglerCheck) error {
```
Comment (noxiouz) [inline senders/juggler/api.go]: Никогда не возвращает error - нужен?

File: senders/juggler/api.go [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+	logger.Infof("%s Update check json payload: %s", js.id, cJSON)
+
+	errs := make(map[string]string, 0)
+	for _, host := range js.JHosts {
+		url := fmt.Sprintf(updateCheckURL, host)
+		resp, err := httpclient.Post(ctx, url, "application/json", bytes.NewReader(cJSON))
+		switch err {
+		case nil:
+			defer resp.Body.Close()
+			body, err := ioutil.ReadAll(resp.Body)
```
Comment (noxiouz) [inline senders/juggler/api.go]: resp.Body.Close()
чтобы дефер не ждать

File: senders/juggler/plugins.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+			if v, err := toLuaValue(l, item, dumper); err == nil {
+				inTable.RawSetString(rv.Type().Field(i).Name, v)
+			} else {
+				return nil, err
+			}
+		}
+		return inTable, nil
+	default:
+		v, err := dumper(rv)
+		if err == nil {
```
Comment (noxiouz) [inline senders/juggler/plugins.go]: if err != nil {
    return nil, err
}
обычно так

File: senders/agave/agave.go [hunk 5]
```diff
@@ -57,58 +50,102 @@ type AgaveConfig struct {
...
+
+	var wg sync.WaitGroup
 	for subgroup, value := range repacked {
-		go as.handleOneItem(subgroup, strings.Join(value, "+"))
+		wg.Add(1)
+		go as.handleOneItem(ctx, subgroup, strings.Join(value, "+"), &wg, e)
 	}
 
+	go func() {
+		wg.Wait()
```
Comment (noxiouz) [inline senders/agave/agave.go]: can we just wait w\o go?

File: senders/agave/agave.go [hunk 6]
```diff
@@ -119,89 +156,101 @@ func (as *AgaveSender) send(data tasks.DataType) (map[string][]string, error) {
...
-		logger.Debugf("%s %s", as.Id, req.URL)
-		resp, err := AgaveHttpClient.Do(req)
-		if err != nil {
-			logger.Errf("%s Unable to do request %s", as.Id, err)
-			continue
-		}
-		defer resp.Body.Close()
+		logger.Debugf("%s %s", as.id, req.URL)
+		resp, err := httpclient.Do(ctx, req)
+		///
```
Comment (noxiouz) [inline senders/agave/agave.go]: ///  ?

File: senders/agave/agave.go [hunk 7]
```diff
@@ -119,89 +156,101 @@ func (as *AgaveSender) send(data tasks.DataType) (map[string][]string, error) {
...
+		switch err {
+		case nil:
+			defer resp.Body.Close()
+			body, err := ioutil.ReadAll(resp.Body)
```
Comment (noxiouz) [inline senders/agave/agave.go]: reps.Body.Close

File: senders/juggler/api.go [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+		//do=1&include_children=true&tag_name=combaine&host_name=
+		query.Set("host_name", js.Host)
+		url := fmt.Sprintf(getChecksURL, jhost, query.Encode())
+		logger.Infof("%s Query check %s", js.id, url)
+
+		resp, err := httpclient.Get(ctx, url)
+		switch err {
+		case nil:
+			defer resp.Body.Close()
+			body, rerr := ioutil.ReadAll(resp.Body)
```
Comment (noxiouz) [inline senders/juggler/api.go]: resp.Body.Close()

File: senders/juggler/config.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+	var path = os.Getenv("JUGGLER_CONFIG")
+	if len(path) == 0 {
+		path = defaultConfigPath
+	}
+
+	rawConfig, err := ioutil.ReadFile(path)
+	if err != nil {
+		return
+	}
+	err = yaml.Unmarshal(rawConfig, &conf)
```
Comment (noxiouz) [inline senders/juggler/config.go]: check err

File: senders/juggler/juggler.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	}
+	return nil
+}
+
+func (js *Sender) sendInternal(ctx context.Context, events []jugglerEvent) error {
+	var jWg sync.WaitGroup
+	var sendEeventsFailed int32
+	for _, jFront := range js.Config.JFrontend {
+		jWg.Add(1)
+		go func(front string, jEs []jugglerEvent, wg *sync.WaitGroup) {
```
Comment (noxiouz) [inline senders/juggler/juggler.go]: jWg is accessabel via closure. No need to pass it.

File: senders/juggler/juggler.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+		Config: conf,
+		id:     id,
+		state:  nil,
+	}, nil
+}
+
+// Send make all things abount juggler sender tasks
+func (js *Sender) Send(ctx context.Context, data []tasks.AggregationResult) error {
+	logger.Debugf("%s Load lua plugin %s", js.id, js.Plugin)
+	state, err := LoadPlugin(js.PluginsDir, js.Plugin)
```
Comment (noxiouz) [inline senders/juggler/juggler.go]: defer state.Close()?

File: senders/solomon/solomon.go [hunk 12]
```diff
@@ -228,33 +238,27 @@ func (s *solomonClient) Send(task tasks.DataType, timestamp uint64) error {
...
-		req.Header.Add("Content-Type", "application/json")
-
-		logger.Debugf("%s attempting to send. Worker %d. Attempt %d", job.SolCli.id, w.Id, attempt)
-		ctx, cancelFunc := context.WithTimeout(context.TODO(), time.Duration(job.SolCli.Timeout)*time.Millisecond)
-		defer cancelFunc()
-		resp, err := httpclient.Do(ctx, req)
+		logger.Debugf("%s attempting to send. Worker %d. Attempt %d", job.SolCli.id, w.id, attempt)
+		ctx, cancel := context.WithTimeout(context.TODO(), time.Duration(job.SolCli.Timeout)*time.Millisecond)
+		defer cancel()
+		resp, err := httpclient.Post(ctx, job.SolCli.API, "application/json", bytes.NewReader(job.PushData))
```
Comment (noxiouz) [inline senders/solomon/solomon.go]: just cancel() after Post

File: senders/juggler/config.go (hunk 9 above)
Comment (sakateka) [inline senders/juggler/config.go]: done

File: senders/juggler/api.go [hunk 13]
```diff
@@ -138,26 +147,90 @@ func (js *Sender) ensureCheck(ctx context.Context, hostChecks jugglerResponse, t
...
+					js.Methods = strings.Split(js.Method, ",")
+				} else {
+					js.Methods = []string{"GOLEM"}
+				}
+			}
+			methodsOutdated := false
+			if len(check.Methods) != len(js.Methods) {
+				methodsOutdated = true
+			} else {
+				checkMSet := make(map[string]struct{})
```
Comment (noxiouz) [inline senders/juggler/api.go]: make(T, len(Methods))

File: senders/juggler/api.go [hunk 14]
```diff
@@ -138,26 +147,90 @@ func (js *Sender) ensureCheck(ctx context.Context, hostChecks jugglerResponse, t
 	for _, t := range triggers {
 		check, ok := services[t.Service]
 		if !ok {
+			logger.Infof("%s Add new check %s.%s", js.id, js.Host, t.Service)
```
Comment (noxiouz) [inline senders/juggler/api.go]: Too complex function. May be split?

**Pull request merged**